[Style] Convert the 'font-family' property to strong style types

weinig opened the pull request:
#### 29fea9f69f1c02064656297f999e2fe4d87eac12
<pre>
[Style] Convert the &apos;font-family&apos; property to strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=302279">https://bugs.webkit.org/show_bug.cgi?id=302279</a>

Reviewed by NOBODY

Converts the &apos;font-family&apos; property to use strong style types.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSFontValue.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleBuilderStateInlines.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/fonts/StyleFontFamily.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontFamily.h: Added.
* Source/WebCore/style/values/fonts/StyleFontFamilyName.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontFamilyName.h: Added.

Canonical link: <a href="https://commits.webkit.org/303001@main">https://commits.webkit.org/303001@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/40d6d4cedf16b9ede9b828becc6ff107af494d1e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130847 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3170 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41806 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138275 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82505 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132718 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3152 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3013 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99702 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/67514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133793 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2270 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117170 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80396 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2190 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35298 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81529 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110783 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35801 "Found 1 new test failure: fast/files/blob-stream-error.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140752 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2914 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2628 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108224 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2960 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113501 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108144 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27521 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2232 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31927 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55948 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2982 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66374 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2803 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3003 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2911 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -230,6 +231,7 @@ inline FlexDirection RenderStyle::flexDirection() const { return static_cast<Fle
 inline Style::FlexGrow RenderStyle::flexGrow() const { return m_nonInheritedData->miscData->flexibleBox->flexGrow; }
 inline Style::FlexShrink RenderStyle::flexShrink() const { return m_nonInheritedData->miscData->flexibleBox->flexShrink; }
 inline FlexWrap RenderStyle::flexWrap() const { return static_cast<FlexWrap>(m_nonInheritedData->miscData->flexibleBox->flexWrap); }
+inline Style::FontFamilies RenderStyle::fontFamily() const { return Style::FontFamilies { fontDescription().families(), fontDescription().isSpecifiedFont() }; }
```
```suggestion
inline Style::FontFamilies RenderStyle::fontFamily() const { return { fontDescription().families(), fontDescription().isSpecifiedFont() }; }
```

darinadler commented on Source/WebCore/style/values/fonts/StyleFontFamily.cpp:
```diff
@@ -0,0 +1,120 @@
+
+        if (isSystemFontShorthand(valueID)) {
+            auto fontShorthand = lowerFontShorthand(valueID);
+            auto family = SystemFontDatabase::singleton().systemFontShorthandFamily(fontShorthand);
```
```suggestion
            auto family = SystemFontDatabase::singleton().systemFontShorthandFamily(lowerFontShorthand(valueID));
```

darinadler commented on Source/WebCore/style/values/fonts/StyleFontFamily.cpp:
```diff
@@ -0,0 +1,120 @@
+
+        if (auto family = genericFontFamily(valueID); !family.isNull()) {
+            return {
+                AtomString { WTFMove(family) },
```
```suggestion
                WTFMove(family),
```

darinadler commented on Source/WebCore/style/values/fonts/StyleFontFamily.cpp:
```diff
@@ -0,0 +1,120 @@
+            ASSERT(!family.isEmpty());
+
+            return {
+                AtomString { WTFMove(family) },
```
```suggestion
                WTFMove(family),
```

darinadler commented on Source/WebCore/style/values/fonts/StyleFontFamily.h:
```diff
@@ -0,0 +1,148 @@
+
+        value_type operator*() const { return SingleFontFamily { *m_it }; }
+
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+        const_iterator& operator++() { ++m_it; return *this; }
+        const_iterator operator++(int) { auto result = *this; ++(*this); return result; }
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
```
Like you, I am surprised that we need to use the `UNSAFE` macros here; aren’t the operators on the underlying iterators already safe, and if so, why is this code unsafe? Or if that line of thinking is wrong, maybe we should can do this based on index-style iterators rather than pointer-style iterators, although that requires holding some kind of pointer to the `RefCountedFixedVector`.

weinig committed:
[Style] Convert the 'font-family' property to strong style types
https://bugs.webkit.org/show_bug.cgi?id=302279

Reviewed by NOBODY

Converts the 'font-family' property to use strong style types.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSFontValue.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleBuilderStateInlines.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/fonts/StyleFontFamily.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontFamily.h: Added.
* Source/WebCore/style/values/fonts/StyleFontFamilyName.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontFamilyName.h: Added.

Canonical link: https://commits.webkit.org/303001@main
